lib/models/patch/file-patch.js

```diff
@@ -0,0 +1,283 @@
+      return false;
+    }
+
+    return this.oldFile.isExecutable() && !this.newFile.isExecutable() ||
+      !this.oldFile.isExecutable() && this.newFile.isExecutable();
+  }
+
+  hasSymlink() {
+    return Boolean(this.getOldFile().getSymlink() || this.getNewFile().getSymlink());
+  }
+
+  hasTypechange() {
+    if (!this.oldFile.isPresent() || !this.newFile.isPresent()) {
+      return false;
+    }
+
+    return this.oldFile.isSymlink() && !this.newFile.isSymlink() ||
+      !this.oldFile.isSymlink() && this.newFile.isSymlink();
+  }
+
+  getPath() {
+    return this.getOldPath() || this.getNewPath();
+  }
+
+  getStatus() {
+    return this.getPatch().getStatus();
+  }
+
+  getHunks() {
+    return this.getPatch().getHunks();
+  }
+
+  clone(opts = {}) {
+    return new this.constructor(
+      opts.oldFile !== undefined ? opts.oldFile : this.oldFile,
+      opts.newFile !== undefined ? opts.newFile : this.newFile,
+      opts.patch !== undefined ? opts.patch : this.patch,
+    );
+  }
+
+  getStagePatchForLines(selectedLineSet) {
+    if (this.patch.getChangedLineCount() === selectedLineSet.size) {
+      if (this.hasTypechange() && this.getStatus() === 'deleted') {
+        // handle special case when symlink is created where a file was deleted. In order to stage the file deletion,
+        // we must ensure that the created file patch has no new file
+        return this.clone({newFile: nullFile});
+      } else {
+        return this;
+      }
+    } else {
+      const patch = this.patch.getStagePatchForLines(selectedLineSet);
+      if (this.getStatus() === 'deleted') {
+        // Populate newFile
+        return this.clone({newFile: this.getOldFile(), patch});
+      } else {
+        return this.clone({patch});
+      }
+    }
+  }
+
+  getStagePatchForHunk(selectedHunk) {
+    return this.getStagePatchForLines(new Set(selectedHunk.getBufferRows()));
+  }
+
+  getUnstagePatchForLines(selectedLineSet) {
+    if (this.patch.getChangedLineCount() === selectedLineSet.size) {
+      const patch = this.patch.getFullUnstagedPatch();
+      if (this.hasTypechange() && this.getStatus() === 'added') {
+        // handle special case when a file was created after a symlink was deleted.
+        // In order to unstage the file creation, we must ensure that the unstage patch has no new file,
+        // so when the patch is applied to the index, there file will be removed from the index.
+        return this.clone({oldFile: nullFile, patch});
+      } else {
+        return this.clone({patch});
+      }
+    } else {
+      const patch = this.patch.getUnstagePatchForLines(selectedLineSet);
+      if (this.getStatus() === 'added') {
+        return this.clone({oldFile: this.getNewFile(), patch});
+      } else {
+        return this.clone({patch});
+      }
+    }
+  }
+
+  getUnstagePatchForHunk(hunk) {
+    return this.getUnstagePatchForLines(new Set(hunk.getBufferRows()));
+  }
+
+  getFirstChangeRange() {
```

Comment:
Curious what the purpose of this method is. There are tests for it, but it's unclear why it's necessary or how it's meant to be used.